In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
import seaborn as sns
import numpy as np

In [3]:
import warnings
warnings.filterwarnings('ignore')

In [4]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/GaBP2/GaBP2_2024/3_Class_SMOTE_Data_SISSO/SISSO_data_Class.csv')
df.head()

,compound,BandGap_class,Volume,vand_rad_Avg,vand_rad_Min,vand_rad_Max,cov_rad_Avg,cov_rad_Min,cov_rad_Max,atom_rad_Avg,atom_rad_Min,atom_rad_Max,density_Avg,density_Min,density_Max
0,ZnGeN2,0,181.39,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13
1,ZnGeN2,1,181.83,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13
2,ZnGeN2,0,181.56,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13
3,AlGaN2,1,178.46,175.33,155,187,104.67,71,122,106.67,65,130,3.14,0.81,5.91
4,AlGaN2,1,178.36,175.33,155,187,104.67,71,122,106.67,65,130,3.14,0.81,5.91


In [5]:
df['Volume_density_Avg']= np.power((df['Volume']*df['density_Avg']),1/4)
df['Volume_density_Avg']= df['Volume_density_Avg'].round(decimals = 2)
df['Volume_atom_rad_Avg']= np.power((df['Volume']*df['atom_rad_Avg']),1/4)
df['Volume_atom_rad_Avg'] = df['Volume_atom_rad_Avg'].round(decimals = 2)
df.head()

,compound,BandGap_class,Volume,vand_rad_Avg,vand_rad_Min,vand_rad_Max,cov_rad_Avg,cov_rad_Min,cov_rad_Max,atom_rad_Avg,atom_rad_Min,atom_rad_Max,density_Avg,density_Min,density_Max,Volume_density_Avg,Volume_atom_rad_Avg
0,ZnGeN2,0,181.39,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13,5.32,11.84
1,ZnGeN2,1,181.83,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13,5.32,11.85
2,ZnGeN2,0,181.56,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13,5.32,11.84
3,AlGaN2,1,178.46,175.33,155,187,104.67,71,122,106.67,65,130,3.14,0.81,5.91,4.87,11.75
4,AlGaN2,1,178.36,175.33,155,187,104.67,71,122,106.67,65,130,3.14,0.81,5.91,4.86,11.74


In [6]:
df.shape

(144, 17)

In [7]:
df_comment = df[['Volume_density_Avg', 'Volume_atom_rad_Avg','BandGap_class']]
df_comment.head()

,Volume_density_Avg,Volume_atom_rad_Avg,BandGap_class
0,5.32,11.84,0
1,5.32,11.85,1
2,5.32,11.84,0
3,4.87,11.75,1
4,4.86,11.74,1


In [8]:
df_comment1= df_comment[df_comment['Volume_density_Avg'] <= 5.5]
df_comment1.head()


,Volume_density_Avg,Volume_atom_rad_Avg,BandGap_class
0,5.32,11.84,0
1,5.32,11.85,1
2,5.32,11.84,0
3,4.87,11.75,1
4,4.86,11.74,1


# `Hypothesis Testing After Data Synthesized`

In [9]:
from scipy import stats

bandgap_1 = df[df['BandGap_class'] == 1]
bandgap_0 = df[df['BandGap_class'] == 0]

# Mean and standard deviation
print("Mean and Std for Bandgap = 1")
print(df[['Volume', 'density_Avg', 'atom_rad_Avg']].mean(), df[['Volume', 'density_Avg', 'atom_rad_Avg']].std())

print("Mean and Std for Bandgap = 0")
print(df[['Volume', 'density_Avg', 'atom_rad_Avg']].mean(), df[['Volume', 'density_Avg', 'atom_rad_Avg']].std())

# t-test for Volume * density_Avg
t_stat, p_val = stats.ttest_ind(bandgap_1['Volume'] * bandgap_1['density_Avg'], bandgap_0['Volume'] * bandgap_0['density_Avg'])
print(f'T-test for Volume * density_Avg: t_stat = {t_stat}, p_val = {p_val}')

# t-test for Volume * atom_rad_Avg
t_stat, p_val = stats.ttest_ind(bandgap_1['Volume'] * bandgap_1['atom_rad_Avg'], bandgap_0['Volume'] * bandgap_0['atom_rad_Avg'])
print(f'T-test for Volume * atom_rad_Avg: t_stat = {t_stat}, p_val = {p_val}')


Mean and Std for Bandgap = 1
Volume          256.058099
density_Avg       4.348855
atom_rad_Avg    115.611231
dtype: float64 Volume          95.607059
density_Avg      1.027227
atom_rad_Avg     9.958562
dtype: float64
Mean and Std for Bandgap = 0
Volume          256.058099
density_Avg       4.348855
atom_rad_Avg    115.611231
dtype: float64 Volume          95.607059
density_Avg      1.027227
atom_rad_Avg     9.958562
dtype: float64
T-test for Volume * density_Avg: t_stat = -14.118456605134813, p_val = 7.883531279971937e-29
T-test for Volume * atom_rad_Avg: t_stat = -15.629622987637985, p_val = 1.1609327950005447e-32


## For refrence

In [10]:
# Mean and Std for Bandgap = 1
# Volume          256.058099
# density_Avg       4.348855
# atom_rad_Avg    115.611231
# dtype: float64 Volume          95.607059
# density_Avg      1.027227
# atom_rad_Avg     9.958562
# dtype: float64
# Mean and Std for Bandgap = 0
# Volume          256.058099
# density_Avg       4.348855
# atom_rad_Avg    115.611231
# dtype: float64 Volume          95.607059
# density_Avg      1.027227
# atom_rad_Avg     9.958562
# dtype: float64
# T-test for Volume * density_Avg: t_stat = -14.118456605134813, p_val = 7.883531279971937e-29
# T-test for Volume * atom_rad_Avg: t_stat = -15.629622987637985, p_val = 1.1609327950005447e-32

# `Hypothesis Testing Actual Data`

In [11]:
Actual_Data= pd.read_csv("/content/drive/MyDrive/Colab Notebooks/GaBP2/GaBP2_2024/2_SMOTE_Data/Descriptor_GaBP2__20_Sep_24.csv")
Actual_Data.head(2)

,compound,BandGap,Volume,vand_rad_Avg,vand_rad_Min,vand_rad_Max,cov_rad_Avg,cov_rad_Min,cov_rad_Max,atom_rad_Avg,atom_rad_Min,atom_rad_Max,density_Avg,density_Min,density_Max
0,ZnGeN2,2.44,181.39,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13
1,ZnGeN2,3.02,181.83,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13


#### `Removed metalic bandgap`

In [12]:
Actual_Data = Actual_Data[Actual_Data['BandGap'].between(0.1, 5)]
Actual_Data.head(2)

,compound,BandGap,Volume,vand_rad_Avg,vand_rad_Min,vand_rad_Max,cov_rad_Avg,cov_rad_Min,cov_rad_Max,atom_rad_Avg,atom_rad_Min,atom_rad_Max,density_Avg,density_Min,density_Max
0,ZnGeN2,2.44,181.39,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13
1,ZnGeN2,3.02,181.83,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13


In [13]:
Actual_Data['BandGap_class']= Actual_Data['BandGap'].apply(lambda z: 0 if z <= 3 else 1)
Actual_Data['BandGap_class'].head()

,BandGap_class
0,0
1,1
2,0
3,1
4,1


In [14]:
pd.value_counts(Actual_Data['BandGap_class'])

,count
BandGap_class,
0,72
1,7


# `Design Feature via SISSO`

In [15]:
Actual_Data['Volume_density_Avg_cbrt']= (np.cbrt(Actual_Data['Volume']*Actual_Data['density_Avg'])).round(decimals = 2)
Actual_Data['Volume_atom_rad_Avg_cbrt']= (np.cbrt(Actual_Data['Volume']*Actual_Data['atom_rad_Avg'])).round(decimals = 2)
Actual_Data.head()

,compound,BandGap,Volume,vand_rad_Avg,vand_rad_Min,vand_rad_Max,cov_rad_Avg,cov_rad_Min,cov_rad_Max,atom_rad_Avg,atom_rad_Min,atom_rad_Max,density_Avg,density_Min,density_Max,BandGap_class,Volume_density_Avg_cbrt,Volume_atom_rad_Avg_cbrt
0,ZnGeN2,2.44,181.39,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13,0,9.29,26.98
1,ZnGeN2,3.02,181.83,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13,1,9.30,27.01
2,ZnGeN2,2.73,181.56,168.33,139,211,104.33,71,122,108.33,65,135,4.42,0.81,7.13,0,9.29,26.99
3,AlGaN2,4.59,178.46,175.33,155,187,104.67,71,122,106.67,65,130,3.14,0.81,5.91,1,8.24,26.70
4,AlGaN2,4.63,178.36,175.33,155,187,104.67,71,122,106.67,65,130,3.14,0.81,5.91,1,8.24,26.70


# `Use Logistic Regression Formula`

In [16]:
Actual_Data1= Actual_Data[['compound', 'BandGap','BandGap_class', 'Volume',	'density_Avg','atom_rad_Avg','Volume_density_Avg_cbrt', 'Volume_atom_rad_Avg_cbrt']]
Actual_Data1.head(2)

,compound,BandGap,BandGap_class,Volume,density_Avg,atom_rad_Avg,Volume_density_Avg_cbrt,Volume_atom_rad_Avg_cbrt
0,ZnGeN2,2.44,0,181.39,4.42,108.33,9.29,26.98
1,ZnGeN2,3.02,1,181.83,4.42,108.33,9.30,27.01


In [17]:
Actual_Data1.shape

(79, 8)

In [18]:
import numpy as np

# Define the logistic regression parameters
intercept = 59.7791
# Replace X1 and X2 with the corresponding coefficients for:
# c1 = 'Volume_density_Avg' and c2 = 'Volume_atom_rad_Avg'
lasso_coefficients = {
    'Volume_density_Avg_cbrt': -1.3922,
    'Volume_atom_rad_Avg_cbrt': -1.6524
}

# Compute logit(P) for each row in ML_data2
def classify(row):
    # Start with the intercept
    logit_P = intercept

    # Add the weighted contribution from each feature
    logit_P += lasso_coefficients['Volume_density_Avg_cbrt'] * row['Volume_density_Avg_cbrt']
    logit_P += lasso_coefficients['Volume_atom_rad_Avg_cbrt'] * row['Volume_atom_rad_Avg_cbrt']

    # Convert logit to probability using the sigmoid function
    probability = 1 / (1 + np.exp(-logit_P))

    # Assign class based on probability threshold of 0.5
    return 1 if probability > 0.5 else 0

# Example usage:
# Assume X1 is your DataFrame containing 'Volume_density_Avg' and 'Volume_atom_rad_Avg'
Actual_Data1['Predicted_Class'] = Actual_Data1.apply(classify, axis=1)

In [19]:
Actual_Data1.head()

,compound,BandGap,BandGap_class,Volume,density_Avg,atom_rad_Avg,Volume_density_Avg_cbrt,Volume_atom_rad_Avg_cbrt,Predicted_Class
0,ZnGeN2,2.44,0,181.39,4.42,108.33,9.29,26.98,1
1,ZnGeN2,3.02,1,181.83,4.42,108.33,9.30,27.01,1
2,ZnGeN2,2.73,0,181.56,4.42,108.33,9.29,26.99,1
3,AlGaN2,4.59,1,178.46,3.14,106.67,8.24,26.70,1
4,AlGaN2,4.63,1,178.36,3.14,106.67,8.24,26.70,1


In [20]:
from scipy import stats

bandgap_1 = Actual_Data1[Actual_Data1['Predicted_Class'] == 1]
bandgap_0 = Actual_Data1[Actual_Data1['Predicted_Class'] == 0]

# Mean and standard deviation
print("Mean and Std for Bandgap = 1")
print(bandgap_1[['Volume', 'density_Avg', 'atom_rad_Avg']].mean())
print(bandgap_1[['Volume', 'density_Avg', 'atom_rad_Avg']].std())

print("\nMean and Std for Bandgap = 0")
print(bandgap_0[['Volume', 'density_Avg', 'atom_rad_Avg']].mean())
print(bandgap_0[['Volume', 'density_Avg', 'atom_rad_Avg']].std())

# t-test for Volume * density_Avg
t_stat, p_val = stats.ttest_ind(bandgap_1['Volume'] * bandgap_1['density_Avg'], bandgap_0['Volume'] * bandgap_0['density_Avg'])
print(f'\nT-test for Volume * density_Avg: t_stat = {t_stat}, p_val = {p_val}')

# t-test for Volume * atom_rad_Avg
t_stat, p_val = stats.ttest_ind(bandgap_1['Volume'] * bandgap_1['atom_rad_Avg'], bandgap_0['Volume'] * bandgap_0['atom_rad_Avg'])
print(f'T-test for Volume * atom_rad_Avg: t_stat = {t_stat}, p_val = {p_val}')


Mean and Std for Bandgap = 1
Volume          180.955
density_Avg       3.789
atom_rad_Avg    107.833
dtype: float64
Volume          7.244200
density_Avg     0.543025
atom_rad_Avg    2.086023
dtype: float64

Mean and Std for Bandgap = 0
Volume          339.220290
density_Avg       5.124783
atom_rad_Avg    124.565942
dtype: float64
Volume          75.554371
density_Avg      0.912908
atom_rad_Avg     6.993672
dtype: float64

T-test for Volume * density_Avg: t_stat = -5.442422052585593, p_val = 6.02874363609822e-07
T-test for Volume * atom_rad_Avg: t_stat = -6.314389824801235, p_val = 1.5990678687838118e-08


In [21]:
# Mean and Std for Bandgap = 1
# Volume          180.955
# density_Avg       3.789
# atom_rad_Avg    107.833
# dtype: float64
# Volume          7.244200
# density_Avg     0.543025
# atom_rad_Avg    2.086023
# dtype: float64

# Mean and Std for Bandgap = 0
# Volume          339.220290
# density_Avg       5.124783
# atom_rad_Avg    124.565942
# dtype: float64
# Volume          75.554371
# density_Avg      0.912908
# atom_rad_Avg     6.993672
# dtype: float64

# T-test for Volume * density_Avg: t_stat = -5.442422052585593, p_val = 6.02874363609822e-07
# T-test for Volume * atom_rad_Avg: t_stat = -6.314389824801235, p_val = 1.5990678687838118e-08

# `T-Test Results`
Volume × Density (p = 6.03e-07): The result is highly significant. The negative t-statistic (-5.44) confirms that this combined feature is substantially lower in bandgap materials. The difference is not due to chance.
Volume × Atomic Radius (p = 1.60e-08): This is even more significant. With a t-statistic of -6.31, atomic radius combined with volume is the stronger separator between the two classes. Bandgap materials have a much smaller volume-radius product.